certbot/tests/util.py

```diff
@@ -208,21 +208,36 @@ def __getattribute__(self, name):
             return getattr(object.__getattribute__(self, '_mock'), name)
 
     def __setattr__(self, name, value):
+        """ Set attributes on the underlying _mock if the FreezableMock is frozen.
+
+        In cases of return_value and side_effect, these attributes are always passed
+        through to the instance's _mock. return_value cannot be set after the FreezableMock
+        instance has been frozen.
+
+        """
+        if name == 'return_value': # Rationale: __setattr__ takes precedence over properties
+            if self._frozen:
+                msg = ("Changing the return_value of a frozen FreezableMock is forbidden "
+                        "because that would nullify callbacks important to thorough tests.")
+                raise AttributeError(msg)
+            else:
+                return setattr(self._mock, name, value)
+        if name == 'side_effect':
+            return setattr(self._mock, name, value)
         if self._frozen:
             return setattr(self._mock, name, value)
         elif name != '_frozen_set':
             self._frozen_set.add(name)
         return object.__setattr__(self, name, value)
 
-
 def _create_get_utility_mock():
     display = FreezableMock()
     for name in interfaces.IDisplay.names():  # pylint: disable=no-member
         if name != 'notification':
-            frozen_mock = FreezableMock(frozen=True, func=_assert_valid_call)
+            frozen_mock = FreezableMock(frozen=False, func=_assert_valid_call)
```

Comment:
This mock object is now no longer frozen when we return it. In practice, this probably doesn't matter because the only attribute we defined was `_func` and it'd be very unusual for a user of this function to change this value.

With that said, I see no reason it can't be frozen. What if we changed the definition of `__setattr__` above to something like:
```
def __setattr__(self, name, value):
    """ Set attributes on the underlying _mock if the FreezableMock is frozen.

    In cases of return_value and side_effect, these attributes are always passed
    through to the instance's _mock.

    """
    if self._frozen:
        if name in self._frozen_set:
            raise AttributeError('Cannot change frozen attribute ' + name)
        else:
            return setattr(self._mock, name, value)
   
    if name != '_frozen_set':
        self._frozen_set.add(name)

    if name in ('return_value', 'side_effect',):
        return setattr(self._mock, name, value)
    else:
        return object.__setattr__(self, name, value)
```
With this approach, the way a `FreezableMock` object works is still consistent. We do treat `return_value` and `side_effect` specially, but I personally think this is just making good use of the tools available to us to avoid reinventing the wheel than an ugly hack.

Note: With this approach, the `return_value` parameter to `__init__` would have to be modified slightly. I think this can be done by setting `self.return_value = return_value` before `self._frozen` is defined but after defining `self._mock`.